process/publisherHandler.go

```diff
@@ -0,0 +1,208 @@
+	"fmt"
+	"time"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-ws-connector-firehose-go/data"
+	"github.com/multiversx/mx-chain-ws-connector-firehose-go/data/hyperOutportBlocks"
+)
+
+const minRetryDudrationInMilliseconds = 100
+
+type publisherHandler struct {
+	handler               HyperBlockPublisher
+	outportBlocksPool     BlocksPool
+	dataAggregator        DataAggregator
+	retryDuration         time.Duration
+	firstCommitableBlocks map[uint32]uint64
+
+	blocksChan chan []byte
+	cancelFunc func()
+	closeChan  chan struct{}
+}
+
+// PublisherHandlerArgs defines the argsuments needed to create a new publisher handler
+type PublisherHandlerArgs struct {
+	Handler                     HyperBlockPublisher
+	OutportBlocksPool           BlocksPool
+	DataAggregator              DataAggregator
+	RetryDurationInMilliseconds uint64
+	FirstCommitableBlocks       map[uint32]uint64
+}
+
+// NewPublisherHandler creates a new publisher handler component
+func NewPublisherHandler(args PublisherHandlerArgs) (*publisherHandler, error) {
+	if check.IfNil(args.Handler) {
+		return nil, ErrNilPublisher
+	}
+	if check.IfNil(args.OutportBlocksPool) {
+		return nil, ErrNilBlocksPool
+	}
+	if check.IfNil(args.DataAggregator) {
+		return nil, ErrNilDataAggregator
+	}
+	if args.FirstCommitableBlocks == nil {
+		return nil, ErrNilFirstCommitableBlocks
+	}
+	if args.RetryDurationInMilliseconds < minRetryDudrationInMilliseconds {
+		return nil, fmt.Errorf("%w for retry duration: provided %d, min required %d",
+			ErrInvalidValue, args.RetryDurationInMilliseconds, minRetryDudrationInMilliseconds)
+	}
+
+	ph := &publisherHandler{
+		handler:               args.Handler,
+		outportBlocksPool:     args.OutportBlocksPool,
+		dataAggregator:        args.DataAggregator,
+		retryDuration:         time.Duration(args.RetryDurationInMilliseconds) * time.Millisecond,
+		firstCommitableBlocks: args.FirstCommitableBlocks,
+		blocksChan:            make(chan []byte),
+		closeChan:             make(chan struct{}),
+	}
+
+	var ctx context.Context
+	ctx, ph.cancelFunc = context.WithCancel(context.Background())
+
+	go ph.startListener(ctx)
+
+	return ph, nil
+}
+
+func (ph *publisherHandler) startListener(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("closing publisherHandler listener")
+			return
+		case headerHash := <-ph.blocksChan:
+			ph.handlePublishEvent(headerHash)
+		}
+	}
+}
+
+// PublishBlock will push header hash to blocks channel
+func (ph *publisherHandler) PublishBlock(headerHash []byte) error {
+	select {
+	case ph.blocksChan <- headerHash:
+	case <-ph.closeChan:
+	}
+
+	return nil
+}
+
+func (ph *publisherHandler) handlePublishEvent(headerHash []byte) {
+	for {
+		err := ph.handlerHyperOutportBlock(headerHash)
+		if err == nil {
+			return
+		}
+
+		log.Error("failed to publish hyper block event", "error", err)
+		time.Sleep(ph.retryDuration)
+	}
+}
```

Comment:
i don't think that a ticker pattern matches here very well, it's more like a retrial mechanism, most of the time it will work from the first try, we need to make sure to try again in case of failure; it might be over complication with a ticker

Reply:
ok, but shouldn't be there some `maximumAttempts` or some way to fail if it is taking too long? I don't like the idea of this running idefinetly, but I might be biased. Might be helpful: https://github.com/avast/retry-go

Reply:
our main goal is to avoid losing any event, so we should try and stay blocked until it works, we shouldn't have an alternative